adapters/repos/db/lsmkv/contentReader/content_reader.go

```diff
@@ -0,0 +1,248 @@
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package contentReader
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"os"
+	"syscall"
+
+	lru "github.com/hashicorp/golang-lru/v2"
+
+	"github.com/edsrzf/mmap-go"
+)
+
+const (
+	uint64Len = 8
+	uint32Len = 4
+)
+
+type ContentReader interface {
+	ReadValue(offset uint64) (byte, uint64)
+	ReadRange(offset uint64, length uint64, outBuf []byte) ([]byte, uint64)
+	ReadUint64(offset uint64) (uint64, uint64)
+	ReadUint32(offset uint64) (uint32, uint64)
+	Length() uint64
+	Close() error
+	NewWithOffsetStart(start uint64) (ContentReader, error)
+	NewWithOffsetStartEnd(start uint64, end uint64) (ContentReader, error)
+	ReaderFromOffset(start uint64, end uint64) io.Reader
+}
+
+type MMap struct {
+	contents []byte
+}
+
+func (c MMap) ReadValue(offset uint64) (byte, uint64) {
+	return c.contents[offset], offset + 1
+}
+
+func (c MMap) ReadRange(offset uint64, length uint64, outBuf []byte) ([]byte, uint64) {
+	if outBuf == nil {
+		return c.contents[offset : offset+length], offset + length
+	}
+	copy(outBuf, c.contents[offset:offset+length])
+	return outBuf, offset + length
+}
+
+func (c MMap) ReadUint64(offset uint64) (uint64, uint64) {
+	return binary.LittleEndian.Uint64(c.contents[offset : offset+uint64Len]), offset + uint64Len
+}
+
+func (c MMap) ReadUint32(offset uint64) (uint32, uint64) {
+	return binary.LittleEndian.Uint32(c.contents[offset : offset+uint32Len]), offset + uint32Len
+}
+
+func (c MMap) Length() uint64 {
+	return uint64(len(c.contents))
+}
+
+func (c MMap) Close() error {
+	m := mmap.MMap(c.contents)
+	if err := m.Unmap(); err != nil {
+		return fmt.Errorf("close segment: munmap: %w", err)
+	}
+	return nil
+}
+
+func (c MMap) NewWithOffsetStart(start uint64) (ContentReader, error) {
+	if start > uint64(len(c.contents)) {
+		return nil, fmt.Errorf("start offset %d is greater than the length of the file", start)
+	}
+	return c.NewWithOffsetStartEnd(start, uint64(len(c.contents)))
+}
+
+func (c MMap) NewWithOffsetStartEnd(start uint64, end uint64) (ContentReader, error) {
+	if end > uint64(len(c.contents)) {
+		return nil, fmt.Errorf("end offset %d is greater than the length of the contents %d", end, len(c.contents))
+	}
+	return MMap{contents: c.contents[start:end]}, nil
+}
+
+func (c MMap) ReaderFromOffset(start uint64, end uint64) io.Reader {
+	if end == 0 {
+		return bytes.NewReader(c.contents[start:])
+	}
+	return bytes.NewReader(c.contents[start:end])
+}
+
+func NewMMap(contents []byte) ContentReader {
+	return MMap{contents: contents}
+}
+
+func NewPread(contentFile *os.File, size uint64) ContentReader {
+	pageSize := syscall.Getpagesize()
+	l, _ := lru.New[int, []byte](128)
+
+	return Pread{contentFile: contentFile, size: size, startOffset: 0, endOffset: size, cache: l, pageSize: pageSize}
+}
+
+type Pread struct {
+	contentFile *os.File
+	size        uint64
+	startOffset uint64
+	endOffset   uint64
+	cache       *lru.Cache[int, []byte]
```

Comment:
Quite like the idea of this cache, which should mitigate some of the current Pread performance humps I've observed! This should also mean that the Header for segments is always cached as well, which is related to the above.